doc/refactoring_design.md

```diff
@@ -0,0 +1,481 @@
+	OceanTribe.addServiceTemplate(bytes32 templateId);
+	OceanCuration.addServiceTemplate(bytes32 templateId);
+	...
+};
+
+// instantiate template to create a service agreement (randomly choose a provider of data commons)
+function executeAgreement(bytes32 templateId, bytes signature, address consumer, bytes32[] valueHash, uint256[] timeoutValues) public {
+	...
+	// check consumer has permission to consume the service
+	if (dataset is public and free to access || consumer in Tribe registry and has permission) 
+	...
+};
+
+// fulfill the service agreement (after all conditions are fulfilled)
+function fulfillAgreement(bytes32 serviceId) public noPendingFulfillments(serviceId) returns(bool){
+	...
+	// detect new block interval and distribute tokens in current block reward pool
+	if( find a new block interval) {
+		// distribute block reward tokens 
+		BlockReward. sendBlockReward();  
+		// reset the candidate list	
+		BlockReward.clearCandidateList()
+	}
+	...
+	// add current provider of data commons into candidate list of block reward 
+	BlockReward.addCandidate(address provider);
+	...
+};
+
+// fulfill specific service condition (common interface)
+function fulfillCondition(bytes32 serviceId, bytes4 fingerprint, bytes32 valueHash) public returns(bool);
+
+// calculate the amount of payment released to provider and refunded to consumer
+function calculatePayment(bytes32 serviceId) public returns(bool);
+```
+
+
+### 2.4 Payment
+
+Payment module handls all payment processing: lock payment, release payment and refund payment. It provides query functions for external function to check the payment status of a specific service agreement.
+
+```Solidity
+// lock consumer payment
+function lockPayment(bytes32 serviceId, bytes32 assetId, uint256 price) public returns (bool);
+
+// release consumer payment to provider
+function releasePayment(bytes32 serviceId, bytes32 assetId, uint256 price) public returns (bool);
+
+// refund payment to consumer
+function refundPayment(bytes32 serviceId, bytes32 assetId, uint256 price) public returns (bool);
+```
+
+### 2.5 Access Control
+
+This module grant consumer with the access to assets. Consumer has two options to receive the access token:
+
+* **Secret Store**: provider keeps the access token inside Secret Store (SS) and adds consumer into permission contract. As such, consumer has no direct access to the access token, which is securely stored in SS. 
+
+	Javascript implementation of the parity secret store for use in Ocean is [here](https://github.com/oceanprotocol/secret-store-client-js).
+
+* **OEP10 authorization**: provider delivers the encrypted access token through on-chain smart contract.
+
+```Solidity
+// grant access to the consumer in given service agreement
+function grantAccess(bytes32 serviceId, bytes32 assetId) public returns (bool);
+```
+
+
+### 2.6 Tribe
+
+Each service agreement template has its corresponding `Tribe` including:
```

Comment:
I see we should start with the definition of a tribe. For instance, `Tribe` as a group of actors. They agree to have set of rules in order to govern relationships between actors. SLA Template is a transcript of service conditions (tribe rules) that automate the execution of these rules.